### PR TITLE
Turn on session keys, use SIWE for Ethereum, add server verification, add Canvas node

### PR DESCRIPTION
**Pull request by raykyri**
Supersedes https://github.com/hicommonwealth/commonwealth/pull/2394.

- [x] Starts enforcing that users have a valid session when they post, and sign their actions when they post.
- [x] Adds server-side validation of the full Canvas session + action scheme. We were only validating sessions before (#3297)
- [x] Updates API tests (#3308, #3306)
- [x] Supports multiple session keys per address (#3305)
    - [x] Detects when user reauths using a different session key, and tells them
- [x] Changes Ethereum login from EIP712 to SIWE, for better security usability
- [x] Adds a canvas-node package to the Commonwealth runtime, which can be run using yarn start or yarn dev to start a Canvas node that syncs against the CW API endpoint
- [x] Adds a GET /api/oplog endpoint, which queries for recent signed actions, and exports recent actions
- [ ] ~~Adds a POST /api/oplog endpoint, which allows new action+session tuples to be submitted and executed against CW.~~

## Link to Issue
Closes: #3297, #3306, #3308, #3305, #4314

### The flow
- Login with Ethereum, Cosmos, NEAR, Solana, Substrate, Terra, Evmos, Ronin/Axie Infinity chains (preferably with multiple addresses)
- Join community
- Post a thread, comment, and reaction (you should do it successfully)
- Switch account
- Clear localStorage and refresh the browser (for now session keys do not expire so clearing local storage mimics the expiration)
- Try to post a thread, comment, reaction. You should be prompted with the modal which says that the session key is expired. You should try to sign in again with wallet that holds the address shown in the modal.
- If you try to use different address to reauthenticate, you should see warning modal that informs that you should use different address. 

### UI components
- Check new navbar icons and tooltips
- check new user dropdown (addresses section specifically as it has different states eg. signed out, signed in - selected, signed in - not selected)
- the address indicator in left sidebar should be visible only if user joined the community. It shows currently selected address.

Also prepared a loom video that explains how we can test it 

https://www.loom.com/share/f4140088bbbc4223b20fe5eea2821ec4?sid=5ffc072a-384d-4ede-8d03-227c9f7abad3

### CODE REVIEW

**Comment by raykyri** [comment]
This should be a draft right now...

**Comment by raykyri** [comment]
@jnaviask @MuonShot @zakhap 

This is ready for review now! 

Tested across Eth, Cosmos, Evmos, Solana, Polkadot, Terra, and NEAR. 

There is an outstanding issue on react/master where Polkadot login may not work at times, because profiles are not properly loaded preventing the user from selecting an account to login. Terra desktop Station also doesn't work, but WalletConnect does. At times, logging in twice may be necessary (we should really fix that, I can coordinate to take a look at it next).

File: packages/commonwealth/client/scripts/controllers/app/webWallets/polkadot_web_wallet.ts
Comment (jnaviask) [inline packages/commonwealth/client/scripts/controllers/app/webWallets/polkadot_web_wallet.ts:35]: Still necessary?

File: packages/commonwealth/client/scripts/controllers/app/webWallets/walletconnect_web_wallet.ts
Comment (jnaviask) [inline packages/commonwealth/client/scripts/controllers/app/webWallets/walletconnect_web_wallet.ts:7]: Single quotes.

File: packages/commonwealth/client/scripts/controllers/app/webWallets/metamask_web_wallet.ts
Comment (jnaviask) [inline packages/commonwealth/client/scripts/controllers/app/webWallets/metamask_web_wallet.ts:12]: Single quotes.

File: packages/commonwealth/client/scripts/controllers/server/sessionSigners/cosmos.ts
Comment (jnaviask) [inline packages/commonwealth/client/scripts/controllers/server/sessionSigners/cosmos.ts:19]: Use an absolute import here.

File: packages/commonwealth/client/scripts/helpers/canvas.ts
Comment (jnaviask) [inline packages/commonwealth/client/scripts/helpers/canvas.ts:3]: Single quotes.

File: packages/commonwealth/client/scripts/views/modals/session_signin_modal.tsx
Comment (jnaviask) [inline packages/commonwealth/client/scripts/views/modals/session_signin_modal.tsx:95]: Can we use cw_modal here instead of custom modal code?

File: packages/commonwealth/server/routes/canvas.ts
Comment (jnaviask) [inline packages/commonwealth/server/routes/canvas.ts:63]: TODO here?

File: packages/commonwealth/server/routes/canvas.ts
Comment (jnaviask) [inline packages/commonwealth/server/routes/canvas.ts:12]: Undefined type or actually empty?

File: packages/commonwealth/server/routes/canvas.ts
Comment (jnaviask) [inline packages/commonwealth/server/routes/canvas.ts:49]: Can use `success` helper instead if you want.

File: packages/commonwealth/server/routes/createComment.ts
Comment (jnaviask) [inline packages/commonwealth/server/routes/createComment.ts:194]: Under what circumstances might verification fail? Is there any reason to be worried about users accidentally being unable to post, e.g. if they're magic link users and can't sign messages?

File: packages/commonwealth/server/routes/canvas.ts
Comment (jnaviask) [inline packages/commonwealth/server/routes/canvas.ts:14]: When does this route get hit / for what use case?

File: packages/commonwealth/server/util/verifySessionSignature.ts
Comment (jnaviask) [inline packages/commonwealth/server/util/verifySessionSignature.ts:249]: Semicolons...

**Review by jnaviask — CHANGES_REQUESTED** [review]
Reviewed the code -- seems fine, but left a bunch of minor comments + some questions about exact functionality. Have not yet done any manual QA. Looks like CI is breaking, would like to see that fixed.

Generally very excited for this PR to go out, I'm glad that we're not only decentralizing, but also standardizing with SIWE, ADR036, etc.

File: packages/commonwealth/client/scripts/controllers/app/webWallets/polkadot_web_wallet.ts
Comment (raykyri) [reply packages/commonwealth/client/scripts/controllers/app/webWallets/polkadot_web_wallet.ts]: Removed

File: packages/commonwealth/server/routes/canvas.ts
Comment (raykyri) [reply packages/commonwealth/server/routes/canvas.ts:12]: This should be an empty object for now, to be filled in later.

File: packages/commonwealth/server/routes/canvas.ts
Comment (raykyri) [reply packages/commonwealth/server/routes/canvas.ts]: Updated.

File: packages/commonwealth/server/routes/canvas.ts
Comment (raykyri) [reply packages/commonwealth/server/routes/canvas.ts:17]: The p2p node polls this endpoint for the latest signed actions published from Common.

File: packages/commonwealth/server/routes/createComment.ts
Comment (raykyri) [reply packages/commonwealth/server/routes/createComment.ts:194]: Wallet holding accounts should always be able to post, given that the logic flow generates a signing key for every login method we use. This covers all mainline chains (Eth, Cosmos, Polkadot, Solana, Cosmos variants) but we should test those.

- Near logins generate a signing key, but the signature check isn't fully enforced because NEAR addresses are human-readable names stored in a registry on their chain, not hashed pubkeys.
- Axie/Ronin logins behave similarly.

Can't test Near or Axie, per my message in Slack.

**Comment by raykyri** [comment]
Everything except Cosmos Magic logins working now. NEAR, Polkadot not tested.

- Still need to test Axie Ronin login
- Still need to look into using Cosmos bech32 to auth other Cosmos chains

Going to look at that + write a developer note.

**Comment by raykyri** [comment]
A little bit about what's going on here:

Magic's Cosmos integration is still based on Launchpad, the initial stable version of the Cosmos SDK from 2020. The Launchpad signing utilities, and presumably the way they've integrated the HSM, don't support ADR-036.

Since ADR-036 signatures are essentially signed TXes with zero sequence numbers and fees, we patch around that, to use their Launchpad signer to get something which is almost an ADR-036 signature.

We do this by extending our /cosmosProxy to provide a mocked Launchpad-compatible RPC to Magic Link's iframe, allowing Magic to get the state of the address they're signing with, and the chain info. (This proxy would be necessary with or without signature format changes).

This requires us to adjust our Cosmos session signature format slightly, accommodating a cosmos chain ID which is fetched from the RPC node. But in a later PR we might be able to change this, and have the proxy return `""` for the chain ID instead, which would cause Magic to emit fully ADR-036 compliant messages.

Since this PR is getting pretty big and I would like to get more testing done, I've left that out for now.

**Comment by raykyri** [comment]
Mainly want to hold back on that as the signature formats are in flux because of Magic signing which uses older Cosmos Launchpad signature. Still need to update Canvas to support those formats.

**Comment by dillchen** [comment]
Makes sense.

TLDR non blocking on merge.

Let's add a ticket separately for /post related items. I think we should not wait on external parties to update their signing logic for that to go in (even if is an old version). I hope this make sense

**Comment by raykyri** [comment]
Currently blocked on Magic SSOs not working on my local env, need to get a working local development SSO token so I can test the SSO codepaths and `finish_social_login` path.

File: packages/canvas-node/contract.js
Comment (dillchen) [inline packages/canvas-node/contract.js:47]: Are these the actions that verify.js verifies against?

Comment (dillchen) [reply packages/canvas-node/contract.js:47]: Docs here: https://docs.canvas.xyz/docs/canvas/packages/interfaces#actions

File: packages/commonwealth/shared/canvas/serverVerify.ts
Comment (dillchen) [inline packages/commonwealth/shared/canvas/serverVerify.ts:50]: Is there a reason why I don't already use the types interface from canvas-node/actions here to verify the fields?

**Comment by jnaviask** [comment]
Did the keys I sent not work?

File: packages/commonwealth/shared/canvas/serverVerify.ts
Comment (raykyri) [reply packages/commonwealth/shared/canvas/serverVerify.ts:50]: Not yet, there will have to be another merge here, but I'd rather make that change later as this touches a main logic flow and any further changes increase the risk of breaking something.

File: packages/canvas-node/contract.js
Comment (raykyri) [reply packages/canvas-node/contract.js:47]: Yes, although I'd like to update the Canvas packages again to support the updated Magic/Launchpad signing formats, rather than overloading methods as we do now.

**Comment by raykyri** [comment]
I think the keys work now. Dealing with getting /auth/magic to work again, but please take a look at the create thread issue I mentioned in Slack - I can continue testing with create comment, but the bug suggests a more general issue with thread creation/address selection.

**Comment by raykyri** [comment]
This is ready for review again.

- We now pass the `redirectTo` and `chain` through the finish_social_login flow, which is required for Cosmos chains to sign session keys with their Magic wallet.
- Session keys are now generated using Magic wallets. Note that we don't always set the right active address in the login menu since profiles were merged - this should be fixed!
- Fix: Magic SSO login wasn't handling the "Connect a new address" case correctly if you were already logged in. This PR makes /auth/magic actually link the new address to the current chain.

Tested on an Ethereum-based chain, Osmosis, and another Cosmos-based chain (Stride) with email and SSO logins. Not tested in other environments, need input on which other environments will be supporting SSO.

**Comment by raykyri** [comment]
1) As we fix issues with the current active account, one question we should ask is how addresses should be treated while composing new threads/comments. We've de-emphasized them in the UX, but still need to make sure users sign with *some* wallet associated with their account. Should we just switch users to the address where they have a valid session key? 

2) We should almost certainly be showing more info around each of users' addresses:
- if an address was connected via email or SSO, we should show that as a tag/label in the address selector. otherwise it's difficult to tell your addresses apart
- the address picker that shows up when logging in with keplr/other multi-address wallets, that was inherited from polkadot login flow, needs work. it incorrectly truncates cosmos addresses so it's impossible to tell which address you're importing

File: packages/commonwealth/shared/canvas/serverVerify.ts
Comment (dillchen) [reply packages/commonwealth/shared/canvas/serverVerify.ts:50]: Possible to link / lmk when that gets in

**Comment by dillchen** [comment]
> re: setting the correct active address:
> 
> 
> 
> 1) As we fix issues with the current active account, one question we should ask is how addresses should be treated while composing new threads/comments. We've de-emphasized them in the UX, but still need to make sure users sign with *some* wallet associated with their account. Should we just switch users to the address where they have a valid session key? 
> 
> 
> 
> 2) We should almost certainly be showing more info around each of users' addresses:
> 
> - if an address was connected via email or SSO, we should show that as a tag/label in the address selector. otherwise it's difficult to tell your addresses apart
> 
> - the address picker that shows up when logging in with keplr/other multi-address wallets, that was inherited from polkadot login flow, needs work. it incorrectly truncates cosmos addresses so it's impossible to tell which address you're importing

Excellent point, I have a product specific Notion doc on this, hope to discuss next week

**Review by jnaviask — COMMENTED** [review]
Forgot to leave comment

**Comment by jnaviask** [comment]
Some sort of profile redraw issue after logging in with metamask, could be related to ToS modal. Fine on refresh.

2023-05-31-114354_371x96_scrot

Session signing modal should tell me which address I'm on, because right now I'm confused like why did this appear? And then metamask popped up and I signed with the wrong address, which confused me. Basically need better visual indicators around session keys state.

Can't get social login to work -- logged in with my jake@commonwealth email on dydx, went to osmosis, try to post a thread, keeps prompting me to sign.

File: packages/commonwealth/server/util/cosmosProxy.ts
Comment (raykyri) [reply packages/commonwealth/server/util/cosmosProxy.ts:159]: The unknown part of the response data is the blocknum. We could mock that with zero or a recently seen blocknum, but introducing state seems sketchy here and it's not clear how much of Magic actually relies on that blocknum, or more importantly, might in the future as they get smarter about signing, so I've left this as is for now...

**Comment by raykyri** [comment]
> * Can't get social login to work -- logged in with my jake@commonwealth email on dydx, went to osmosis, try to post a thread, keeps prompting me to sign. Tried again with another email, same issue.

Fixed; finish_magic_login now passes a signed login SessionPayload to the backend, which verifies it and adds the Magic address.

As discussed with Jake, we should follow up to this by generating both Cosmos and Eth addresses in the `/finishsociallogin` page and verify them both when we call `/auth/magic` to log the user in; that way the user should always have eth and cosmos session keys whenever they log in with magic, reducing switching complexity and allowing more data to be signed.

- It is confusing for multiple generated user addresses to show up in the address selector - so we should prioritize labeling those addresses using `WalletId` and perhaps even add another piece of metadata for what email/SSO method was used.
- We can also enhance the wallet revalidation flow by adding that metadata.
- We can also enhance the wallet revalidation flow by helping the user switch addresses / switch login methods to one they're familiar with.

As these are all smaller improvements they should probably go in separate PRs to follow this.

> * Some sort of profile redraw issue after logging in with metamask, could be related to ToS modal. Fine on refresh.
>   2023-05-31-114354_371x96_scrot

Likewise, this should be handled in a separate PR as it's an existing issue.

>   Session signing modal should tell me which address I'm on, because right now I'm confused like why did this appear? And then metamask popped up and I signed with the wrong address, which confused me. Basically need better visual indicators around session keys state.

Handled by showing the address that we want the user to log in from.

**Comment by raykyri** [comment]
@jnaviask This should be good for your review again - I tested with dydx and Osmosis and was able to create threads and comments.

**Comment by jnaviask** [comment]
Getting the following behavior:
- Log into an existing account on /dydx created via social login, which has an eth and a cosmos... address
- Try to create a thread on /osmosis (only cosmos... address present), session modal appears
- I log in with session modal, now I have two addresses. Still doesn't work until I explicitly change from cosmos to osmosis (nowhere would I know to do this step).

So, technically it works, but this UX barrier is so severe that I worry we could consider it site-breaking if we release as-is. Contained in the above is two separate issues:
1. We don't generate the signed Osmosis address at initial login, only at revalidation (or, if we do, it didn't appear).
2. We don't properly switch addresses for users on validation.

File: packages/commonwealth/server/passport/magic.ts
Comment (jnaviask) [inline packages/commonwealth/server/passport/magic.ts]: Remove this block.

File: packages/commonwealth/server/passport/magic.ts
Comment (raykyri) [reply packages/commonwealth/server/passport/magic.ts]: Done

File: packages/commonwealth/server/passport/magic.ts
Comment (raykyri) [reply packages/commonwealth/server/passport/magic.ts]: I think this is OK now because we've removed the old generatedAddresses.push() calls.

**Comment by raykyri** [comment]
QA starting states:

* Starting from a [logged in, not logged in] state
* Starting from this page:
    * On a page without chain (homepage or dashboard)
    * On a page in an Eth-based chain community
    * On a page in the Osmosis community
    * On a page in a Cosmos chain community that's not Osmosis
    * On the /finishsociallogin page
* Log in with:
    * An email
    * An SSO method
    * A wallet

Actions:

* Create a thread
* Create a comment
* Create a reaction

* Clear localStorage and try to create the same item again
* Clear localStorage, switch addresses, and try to create the same item again
* Clear localStorage, go through the initial logins again

Check the active addresses list after each of the login steps.

**Comment by raykyri** [comment]
@jnaviask @meeshlin Handled the design review:

- Implements the new copy.
- Added a "switch address" dropdown, that shows other connected addresses, and displays as unavailable if they do not have a session key available.
- Added a separate "Proceed with current address" screen instead of using our current login (would be difficult to filter that down). 

image

image

**Comment by masvelio** [comment]
@raykyri make sure you run `yarn setup-git-hooks` and  `yarn add-prettier-precommit` locally because I can see that the files are not formatted in proper way

File: packages/commonwealth/client/scripts/controllers/server/sessions.ts
Comment (masvelio) [inline packages/commonwealth/client/scripts/controllers/server/sessions.ts:244]: FYI @raykyri @jnaviask I am working right now on this piece. We have to move it out from controller to some reusable logic (custom hook probably), because
1. we should try to not add new logic to the controllers as we are trying to move away from controllers architecture
2. we should not handle UI from the controllers layer

File: packages/commonwealth/client/scripts/controllers/server/sessions.ts
Comment (raykyri) [reply packages/commonwealth/client/scripts/controllers/server/sessions.ts:276]: Let me push a change to add this to the client (since it is UX impacting) but since we are not verifying deletions on the server (goal for this PR being to get client side content signed) it isn't critical to include.

**Comment by masvelio** [comment]
@jnaviask after master merge, some check are again red. I think we need to leave it as it is and focus on making all checks green right before the merge to the master. Otherwise, we will be wasting time fixing this every time master will be merged

**Comment by sachben91** [comment]
https://www.notion.so/buildcommon/Session-keys-product-testing-313fb2c93d9f47e084320c26717597fb Comments that came up on testing are on this Notion. Not all the items are actionable but dont want them to get lost in a comment thread on the PR hence the Notion. It has also been shared on the group chat that includes everyone working on this

**Review by CowMuon — APPROVED** [review]
This is now passing design review (on Frack) and is ready to go to QA. (This should also eliminate the divergence we're seeing between Frack and Production.)

File: packages/commonwealth/server/passport/magic.ts
Comment (CowMuon) [reply packages/commonwealth/server/passport/magic.ts]: Curious why this comment has not been resolved since Jun 1?

File: packages/commonwealth/server/routes/canvas.ts
Comment (CowMuon) [reply packages/commonwealth/server/routes/canvas.ts]: Merge and ship, apparently. (TBC- this is a sarcastic reference to all the unresolved code comments 4 months later)

File: packages/commonwealth/client/scripts/views/modals/session_signin_modal.tsx
Comment (CowMuon) [reply packages/commonwealth/client/scripts/views/modals/session_signin_modal.tsx]: Good suggestion.

**Pull request merged**
